crates/providers/src/signer.rs

```diff
@@ -0,0 +1,152 @@
+///     .provider(RootProvider::new(transport));
+///
+/// provider.send_transaction(TransactionRequest::default()).await;
+/// # }
+/// ```
+pub struct SignerLayer<S> {
+    signer: S,
+}
+
+impl<S> SignerLayer<S> {
+    /// Creates a new signing layer with the given signer.
+    pub fn new(signer: S) -> Self {
+        Self { signer }
+    }
+}
+
+impl<P, N, T, S> ProviderLayer<P, N, T> for SignerLayer<S>
+where
+    P: Provider<N, T>,
+    N: Network,
+    T: Transport + Clone,
+    S: NetworkSigner<N> + Clone,
+{
+    type Provider = SignerProvider<N, T, P, S>;
+
+    fn layer(&self, inner: P) -> Self::Provider {
+        SignerProvider { inner, signer: self.signer.clone(), _phantom: PhantomData }
+    }
+}
+
+/// A locally-signing provider.
+///
+/// Signs transactions locally using a [`NetworkSigner`]
+///
+/// # Note
+///
+/// You cannot construct this provider directly. Use [`ProviderBuilder`] with a [`SignerLayer`].
+///
+/// [`ProviderBuilder`]: crate::ProviderBuilder
+pub struct SignerProvider<N, T, P, S>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    inner: P,
+    signer: S,
+    _phantom: PhantomData<(N, T)>,
+}
+
+#[async_trait]
+impl<N, T, P, S> Provider<N, T> for SignerProvider<N, T, P, S>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+    S: NetworkSigner<N>,
+{
+    fn client(&self) -> ClientRef<'_, T> {
+        self.inner.client()
+    }
+
+    fn weak_client(&self) -> WeakClient<T> {
+        self.inner.weak_client()
+    }
+
+    async fn new_pending_transaction(&self, tx_hash: B256) -> TransportResult<PendingTransaction> {
+        self.inner.new_pending_transaction(tx_hash).await
+    }
+
+    async fn send_transaction(
+        &self,
+        tx: N::TransactionRequest,
+    ) -> TransportResult<PendingTransaction> {
+        let envelope = tx.build(&self.signer).await.map_err(TransportErrorKind::custom)?;
+        let rlp = envelope.encoded_2718();
+
+        self.inner.send_raw_transaction(&rlp).await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::SignerLayer;
+    use crate::{Provider, ProviderBuilder, RootProvider};
+    use alloy_network::{Ethereum, EthereumSigner};
+    use alloy_node_bindings::Anvil;
+    use alloy_primitives::{address, U256, U64};
+    use alloy_rpc_client::RpcClient;
+    use alloy_rpc_types::TransactionRequest;
+    use alloy_transport_http::Http;
+    use reqwest::Client;
+
+    #[tokio::test]
+    async fn poc() {
+        let anvil = Anvil::new().spawn();
+        let url = anvil.endpoint().parse().unwrap();
+        let http = Http::<Client>::new(url);
+
+        let wallet = alloy_signer::Wallet::from(anvil.keys()[0].clone());
+
+        // can we somehow remove the need for <_, Ethereum>? we NEED to call .network<Ethereum>
+        // note: we need to 1) add <_, Ethereum> 2) layer things, and then 3) call .network before
+        // we can call provider
+        let provider = ProviderBuilder::<_, Ethereum>::new()
```

Comment:
Ethereum::provider -> builder

Reply:
this would cause a circular dependency, so not possible currently. will leave for follow up

Reply:
OK